### ISSUE TITLE
register_full_backward_hook does not consistently fire

### ISSUE THREAD

**Issue by Shoeboxam** [pytorch/pytorch#56380]
## 🐛 Bug
In Pytorch 1.8, `Module.register_backward_hook` was deprecated in favor of `Module.register_full_backward_hook`. Unfortunately, `Module.register_full_backward_hook` doesn't fire on the first module it was registered on.

## To Reproduce
1. Register the full backwards hook on each module.
2. call backwards on the network

```python
import torch
import torch.nn as nn
import torch.nn.functional as F


class TestModule(nn.Module):
    def __init__(self, input_size, output_size):
        super().__init__()
        internal_size = 5
        self.linear1 = nn.Linear(input_size, internal_size)
        self.linear2 = nn.Linear(internal_size, output_size)

    def forward(self, x):
        x = self.linear1(x)
        x = F.relu(x)
        x = self.linear2(x)
        return x


model = TestModule(4, 3)


def print_hook(module: nn.Module, _inputs, _outputs):
    print('hook triggered on', module)


for module in model.modules():
    if isinstance(module, nn.Linear):
        module.register_full_backward_hook(print_hook)
        print('added hook to', module)

x, y = torch.rand(size=(20, 4)), torch.randint(low=0, high=3, size=(20,))
loss = F.cross_entropy(model(x), y)

print('backward started')
loss.backward()
print('backward complete')

```

Prints the following:
```
added hook to Linear(in_features=4, out_features=5, bias=True)
added hook to Linear(in_features=5, out_features=3, bias=True)
backward started
hook triggered on Linear(in_features=5, out_features=3, bias=True)
backward complete
```
As you can see, the backward hook for the first linear layer never fired.

## Expected behavior
I expect every hook I register to fire.

## Environment
```
PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.2.3 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.31.1)
CMake version: version 3.19.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] pytorch-lightning==1.1.2
[pip3] torch==1.8.0
[pip3] torchvision==0.8.2
[conda] Could not collect
```

Thank you!

EDIT: Made the repro more succinct.

cc @ezyang @gchanan @albanD @mruberry @jbschlosser

**Comment by albanD** [comment]
Hi,

The new hook will now properly fire only if the gradients for both the input and output of the Module are computed.
But in your case, the input of the first Module does not require gradients. So it is expected that the hook does not fire (because there are no grad_input to print for that Module).

We should add a warning when the hook sees no input that require gradients so that it is less confusing for users.

**Comment by Shoeboxam** [comment]
@albanD I am surprised this is considered expected behavior. If a module requires a gradient, but none of its dependencies require a gradient, I would expect this hook to fire and leave the input argument as None. For my use case, I only need the output gradient of the module.

**Comment by albanD** [comment]
The temporary fix you can use is to make the input require gradients.

**Comment by Shoeboxam** [comment]
Sure, I appreciate the workaround. Setting this flag leaks out into the user-space of the library I'm developing this for. Do you think the behavior will be adjusted in the future? I'm tempted to continue using the deprecated `Module.register_backward_hook` for now.

**Comment by Shoeboxam** [comment]
This function patches a nn.Module to get the first hook to fire:

```python
def patch_first_hook(model):
    def set_requires_grad(arg):
        if isinstance(arg, torch.Tensor):
            arg.requires_grad = True
        return arg

    old_forward = model.forward

    def forward(*args, **kwargs):
        return old_forward(
            *(set_requires_grad(arg) for arg in args),
            **{kw: set_requires_grad(arg) for kw, arg in kwargs.items()})
    setattr(model, 'forward', forward)
```

**Comment by albanD** [comment]
After looking at it in more details, it can have the expected behavior of returning `None` grad for all the inputs.
The PR linked above should fix your issue :)
